File: reference/r-tests/testthat/test-aoi-biometrics.R
test_that("summarise_gazepoint_aoi_biometrics summarises signals by AOI", {
  dat <- data.frame(
    participant = c("P1", "P1", "P1", "P1", "P2", "P2"),
    MEDIA_ID = c(1, 1, 1, 1, 1, 1),
    AOI = c("claim", "claim", "logo", "logo", "claim", "logo"),
    GSR_US = c(1, 2, 3, 4, 5, 6),
    HR = c(70, 72, 74, 76, 78, 80)
  )

  res <- summarise_gazepoint_aoi_biometrics(
    dat,
    aoi_col = "AOI",
    signal_cols = c("GSR_US", "HR"),
    group_cols = c("participant", "MEDIA_ID"),
    min_rows = 1
  )

  expect_s3_class(res, "gazepoint_aoi_biometrics_summary")
  expect_equal(res$overview$signal_count, 2)
  expect_true(nrow(res$summary) > 0)
  expect_true(all(c("mean_value", "median_value", "signal", "aoi_label") %in% names(res$summary)))
})

test_that("summarise_gazepoint_aoi_biometrics can filter AOI labels", {
  dat <- data.frame(
    participant = "P1",
    AOI = c("claim", "claim", "logo"),
    GSR_US = c(1, 2, 3)
  )

  res <- summarise_gazepoint_aoi_biometrics(
    dat,
    aoi_col = "AOI",
    signal_cols = "GSR_US",
    group_cols = "participant",
    valid_aoi_values = "claim"
  )

  expect_equal(unique(res$summary$aoi_label), "claim")
  expect_equal(res$overview$aoi_count, 1)
})

test_that("summarise_gazepoint_aoi_biometrics warns for low-row summaries", {
  dat <- data.frame(
    participant = "P1",
    AOI = c("claim", "logo"),
    GSR_US = c(1, 2)
  )

  res <- summarise_gazepoint_aoi_biometrics(
    dat,
    aoi_col = "AOI",
    signal_cols = "GSR_US",
    group_cols = "participant",
    min_rows = 2
  )

  expect_equal(res$overview$status, "warn_low_rows_in_some_summaries")
  expect_true(any(res$summary$summary_status == "warn_low_rows"))
})

test_that("summarise_gazepoint_aoi_biometrics handles no retained AOI rows", {
  dat <- data.frame(
    participant = "P1",
    AOI = c(NA, ""),
    GSR_US = c(1, 2)
  )

  res <- summarise_gazepoint_aoi_biometrics(
    dat,
    aoi_col = "AOI",
    signal_cols = "GSR_US",
    group_cols = "participant"
  )

  expect_equal(res$overview$status, "fail_no_aoi_rows")
  expect_equal(nrow(res$summary), 0)
})

test_that("prepare_gazepoint_aoi_biometrics_model_data prepares model data", {
  dat <- data.frame(
    participant = c("P1", "P1", "P2", "P2"),
    AOI = c("claim", "logo", "claim", "logo"),
    GSR_US = c(1, 2, 3, 4)
  )

  summary <- summarise_gazepoint_aoi_biometrics(
    dat,
    aoi_col = "AOI",
    signal_cols = "GSR_US",
    group_cols = "participant"
  )

  model_data <- prepare_gazepoint_aoi_biometrics_model_data(
    summary,
    outcome_col = "mean_value",
    predictor_cols = c("aoi_label", "signal"),
    factor_cols = c("aoi_label", "signal"),
    group_cols = "participant",
    standardise_outcome = TRUE,
    standardise_within = "signal"
  )

  expect_s3_class(model_data, "gazepoint_aoi_biometrics_model_data")
  expect_equal(model_data$overview$status, "aoi_biometrics_model_data_prepared")
  expect_equal(model_data$overview$standardise_within, "signal")
  expect_true("mean_value_z" %in% names(model_data$model_data))
  expect_true(is.factor(model_data$model_data$aoi_label))
})

test_that("prepare_gazepoint_aoi_biometrics_model_data standardises within signal", {
  dat <- data.frame(
    participant = c("P1", "P1", "P2", "P2"),
    AOI = c("claim", "logo", "claim", "logo"),
    GSR_US = c(1, 2, 3, 4),
    HR = c(70, 72, 74, 76)
  )

  summary <- summarise_gazepoint_aoi_biometrics(
    dat,
    aoi_col = "AOI",
    signal_cols = c("GSR_US", "HR"),
    group_cols = "participant"
  )

  model_data <- prepare_gazepoint_aoi_biometrics_model_data(
    summary,
    outcome_col = "mean_value",
    predictor_cols = c("aoi_label", "signal"),
    factor_cols = c("aoi_label", "signal"),
    group_cols = "participant",
    standardise_outcome = TRUE,
    standardise_within = "signal"
  )

  by_signal_mean <- tapply(
    model_data$model_data$mean_value_z,
    model_data$model_data$signal,
    mean,
    na.rm = TRUE
  )

  expect_equal(model_data$overview$standardise_within, "signal")
  expect_true(all(abs(by_signal_mean) < 1e-10))
})

test_that("prepare_gazepoint_aoi_biometrics_model_data can standardise across all rows", {
  dat <- data.frame(
    participant = c("P1", "P1", "P2", "P2"),
    AOI = c("claim", "logo", "claim", "logo"),
    GSR_US = c(1, 2, 3, 4),
    HR = c(70, 72, 74, 76)
  )

  summary <- summarise_gazepoint_aoi_biometrics(
    dat,
    aoi_col = "AOI",
    signal_cols = c("GSR_US", "HR"),
    group_cols = "participant"
  )

  model_data <- prepare_gazepoint_aoi_biometrics_model_data(
    summary,
    outcome_col = "mean_value",
    predictor_cols = c("aoi_label", "signal"),
    factor_cols = c("aoi_label", "signal"),
    group_cols = "participant",
    standardise_outcome = TRUE,
    standardise_within = "all"
  )

  expect_equal(model_data$overview$standardise_within, "all")
  expect_true(abs(mean(model_data$model_data$mean_value_z, na.rm = TRUE)) < 1e-10)
})

test_that("prepare_gazepoint_aoi_biometrics_model_data can filter by min_rows", {
  dat <- data.frame(
    participant = "P1",
    AOI = c("claim", "claim", "logo"),
    GSR_US = c(1, 2, 3)
  )

  summary <- summarise_gazepoint_aoi_biometrics(
    dat,
    aoi_col = "AOI",
    signal_cols = "GSR_US",
    group_cols = "participant"
  )

  model_data <- prepare_gazepoint_aoi_biometrics_model_data(
    summary,
    outcome_col = "mean_value",
    predictor_cols = "aoi_label",
    min_rows = 2
  )

  expect_true(all(model_data$model_data$n_rows >= 2))
})

test_that("plot_gazepoint_aoi_biometrics returns ggplot objects", {
  skip_if_not_installed("ggplot2")

  dat <- data.frame(
    participant = c("P1", "P1", "P2", "P2"),
    AOI = c("claim", "logo", "claim", "logo"),
    GSR_US = c(1, 2, 3, 4),
    HR = c(70, 72, 74, 76)
  )

  summary <- summarise_gazepoint_aoi_biometrics(
    dat,
    aoi_col = "AOI",
    signal_cols = c("GSR_US", "HR"),
    group_cols = "participant"
  )

  p <- plot_gazepoint_aoi_biometrics(
    summary,
    value_col = "mean_value",
    plot_type = "point",
    group_col = "participant"
  )

  expect_s3_class(p, "ggplot")
  expect_true(is.data.frame(attr(p, "plot_data")))
})

test_that("plot_gazepoint_aoi_biometrics supports model-data objects", {
  skip_if_not_installed("ggplot2")

  dat <- data.frame(
    participant = c("P1", "P1", "P2", "P2"),
    AOI = c("claim", "logo", "claim", "logo"),
    GSR_US = c(1, 2, 3, 4)
  )

  summary <- summarise_gazepoint_aoi_biometrics(
    dat,
    aoi_col = "AOI",
    signal_cols = "GSR_US",
    group_cols = "participant"
  )

  model_data <- prepare_gazepoint_aoi_biometrics_model_data(
    summary,
    standardise_outcome = TRUE,
    standardise_within = "signal"
  )

  p <- plot_gazepoint_aoi_biometrics(
    model_data,
    value_col = "mean_value_z",
    plot_type = "boxplot"
  )

  expect_s3_class(p, "ggplot")
})

test_that("AOI-biometric helpers validate inputs", {
  dat <- data.frame(
    AOI = "claim",
    GSR_US = 1
  )

  expect_error(
    summarise_gazepoint_aoi_biometrics(dat, aoi_col = "missing"),
    "`aoi_col`"
  )

  expect_error(
    summarise_gazepoint_aoi_biometrics(dat, signal_cols = "missing"),
    "`signal_cols`"
  )

  expect_error(
    prepare_gazepoint_aoi_biometrics_model_data(dat, outcome_col = "missing"),
    "`outcome_col`"
  )

  expect_error(
    prepare_gazepoint_aoi_biometrics_model_data(
      data.frame(
        aoi_label = "claim",
        mean_value = 1
      ),
      outcome_col = "mean_value",
      standardise_outcome = TRUE,
      standardise_within = "signal"
    ),
    "`standardise_within = \"signal\"` requires a `signal` column"
  )

  testthat::skip_if_not_installed("ggplot2")
  expect_error(
    plot_gazepoint_aoi_biometrics(dat, value_col = "missing"),
    "Required plotting columns"
  )
})

File: tests/test_aoi_biometrics.py
import matplotlib.figure
import numpy as np
import pandas as pd
import pytest
import gpbiometricspy as gp


def _summary_data():
    return pd.DataFrame({'participant':['P1','P1','P1','P1','P2','P2'],'MEDIA_ID':[1]*6,'AOI':['claim','claim','logo','logo','claim','logo'],'GSR_US':[1,2,3,4,5,6],'HR':[70,72,74,76,78,80]})


def test_aoi_biometric_summary_and_filter_low_rows_empty():
    r=gp.summarise_gazepoint_aoi_biometrics(_summary_data(),signal_cols=['GSR_US','HR'],group_cols=['participant','MEDIA_ID'],min_rows=1)
    assert r['overview'].iloc[0].signal_count==2 and len(r['summary'])>0 and {'mean_value','median_value','signal','aoi_label'}.issubset(r['summary'].columns)
    dat=pd.DataFrame({'participant':'P1','AOI':['claim','claim','logo'],'GSR_US':[1,2,3]})
    f=gp.summarise_gazepoint_aoi_biometrics(dat,signal_cols='GSR_US',group_cols='participant',valid_aoi_values='claim')
    assert set(f['summary'].aoi_label)=={'claim'} and f['overview'].iloc[0].aoi_count==1
    low=gp.summarise_gazepoint_aoi_biometrics(pd.DataFrame({'participant':'P1','AOI':['claim','logo'],'GSR_US':[1,2]}),signal_cols='GSR_US',group_cols='participant',min_rows=2)
    assert low['overview'].iloc[0].status=='warn_low_rows_in_some_summaries' and (low['summary'].summary_status=='warn_low_rows').any()
    empty=gp.summarise_gazepoint_aoi_biometrics(pd.DataFrame({'participant':'P1','AOI':[np.nan,''],'GSR_US':[1,2]}),signal_cols='GSR_US',group_cols='participant')
    assert empty['overview'].iloc[0].status=='fail_no_aoi_rows' and empty['summary'].empty


def test_model_data_standardization_modes_and_filter():
    dat=pd.DataFrame({'participant':['P1','P1','P2','P2'],'AOI':['claim','logo','claim','logo'],'GSR_US':[1,2,3,4],'HR':[70,72,74,76]})
    s=gp.summarise_gazepoint_aoi_biometrics(dat,signal_cols=['GSR_US','HR'],group_cols='participant')
    m=gp.prepare_gazepoint_aoi_biometrics_model_data(s,outcome_col='mean_value',predictor_cols=['aoi_label','signal'],factor_cols=['aoi_label','signal'],group_cols='participant',standardise_outcome=True,standardise_within='signal')
    assert m['overview'].iloc[0].status=='aoi_biometrics_model_data_prepared' and m['overview'].iloc[0].standardise_within=='signal'
    assert 'mean_value_z' in m['model_data'] and str(m['model_data'].aoi_label.dtype)=='category'
    means=m['model_data'].groupby('signal',observed=False).mean_value_z.mean(); assert (np.abs(means)<1e-10).all()
    allm=gp.prepare_gazepoint_aoi_biometrics_model_data(s,standardise_outcome=True,standardise_within='all')
    assert abs(allm['model_data'].mean_value_z.mean())<1e-10
    s2=gp.summarise_gazepoint_aoi_biometrics(pd.DataFrame({'participant':'P1','AOI':['claim','claim','logo'],'GSR_US':[1,2,3]}),signal_cols='GSR_US',group_cols='participant')
    filt=gp.prepare_gazepoint_aoi_biometrics_model_data(s2,predictor_cols='aoi_label',min_rows=2)
    assert (filt['model_data'].n_rows>=2).all()


def test_aoi_plot_and_validation():
    dat=pd.DataFrame({'participant':['P1','P1','P2','P2'],'AOI':['claim','logo','claim','logo'],'GSR_US':[1,2,3,4],'HR':[70,72,74,76]})
    s=gp.summarise_gazepoint_aoi_biometrics(dat,signal_cols=['GSR_US','HR'],group_cols='participant')
    fig=gp.plot_gazepoint_aoi_biometrics(s,value_col='mean_value',plot_type='point',group_col='participant')
    assert isinstance(fig,matplotlib.figure.Figure) and isinstance(fig.plot_data,pd.DataFrame)
    m=gp.prepare_gazepoint_aoi_biometrics_model_data(s,standardise_outcome=True,standardise_within='signal')
    fig2=gp.plot_gazepoint_aoi_biometrics(m,value_col='mean_value_z',plot_type='boxplot'); assert isinstance(fig2,matplotlib.figure.Figure)
    with pytest.raises(ValueError,match='aoi_col'): gp.summarise_gazepoint_aoi_biometrics(pd.DataFrame({'AOI':['claim'],'GSR_US':[1]}),aoi_col='missing')
    with pytest.raises(ValueError,match='signal_cols'): gp.summarise_gazepoint_aoi_biometrics(pd.DataFrame({'AOI':['claim'],'GSR_US':[1]}),signal_cols='missing')
    with pytest.raises(ValueError,match='outcome_col'): gp.prepare_gazepoint_aoi_biometrics_model_data(pd.DataFrame({'mean_value':[1]}),outcome_col='missing')
    with pytest.raises(ValueError,match='requires a `signal` column'): gp.prepare_gazepoint_aoi_biometrics_model_data(pd.DataFrame({'aoi_label':['claim'],'mean_value':[1]}),standardise_outcome=True,standardise_within='signal')
    with pytest.raises(ValueError,match='Required plotting columns'): gp.plot_gazepoint_aoi_biometrics(pd.DataFrame({'AOI':['claim'],'GSR_US':[1]}),value_col='missing')
